src/test/java/com/puppycrawl/tools/checkstyle/XpathRegressionTest.java

```diff
@@ -0,0 +1,199 @@
+import com.puppycrawl.tools.checkstyle.xpath.XpathQueryGenerator;
+
+public class XpathRegressionTest extends AbstractModuleTestSupport {
+
+    private static final int DEFAULT_TAB_WIDTH = 4;
+
+    private static final String DELIMITER = " | ";
+
+    private static final String[] NO_VIOLATIONS = CommonUtils.EMPTY_STRING_ARRAY;
+
+    @Rule
+    public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    @Override
+    protected String getPackageLocation() {
+        return "com/puppycrawl/tools/checkstyle/filters/suppressionxpathfilter";
+    }
+
+    private static List<String> generateXpathQueries(File fileToProcess, int violationLineNumber,
+                                              int violationColumnNumber) throws Exception {
+        final FileText fileText = new FileText(fileToProcess,
+                StandardCharsets.UTF_8.name());
+        final DetailAST rootAst = JavaParser.parseFile(fileToProcess,
+                JavaParser.Options.WITH_COMMENTS);
+        final XpathQueryGenerator queryGenerator = new XpathQueryGenerator(rootAst,
+                violationLineNumber, violationColumnNumber, fileText, DEFAULT_TAB_WIDTH);
+        return queryGenerator.generate();
+    }
+
+    private static DefaultConfiguration createCheckerWithTreeWalkerModules(
+            DefaultConfiguration... children) {
+        final DefaultConfiguration treeWalkerConfig = createModuleConfig(TreeWalker.class);
+        for (DefaultConfiguration child : children) {
+            treeWalkerConfig.addChild(child);
+        }
+        return createRootConfig(treeWalkerConfig);
+    }
+
+    private static void verifyXpathQueries(List<String> generatedXpathQueries,
+                                           List<String> expectedXpathQueries) {
+        assertEquals("Generated queries do not match expected ones", expectedXpathQueries,
+                generatedXpathQueries);
+    }
+
+    private String createSuppressionsXpathConfigFile(String checkName,
+                                                     List<String> xpathQueries)
+            throws Exception {
+
+        final File suppressionsXpathConfigFile = temporaryFolder.newFile();
+        try (Writer bw = Files.newBufferedWriter(suppressionsXpathConfigFile.toPath(),
+                StandardCharsets.UTF_8)) {
+            bw.write("<?xml version=\"1.0\"?>\n");
+            bw.write("<!DOCTYPE suppressions PUBLIC\n");
+            bw.write("    \"-//Puppy Crawl//DTD Suppressions Xpath Experimental 1.2//EN\"\n");
+            bw.write("    \"http://checkstyle.sourceforge.net/dtds/");
+            bw.write("suppressions_1_2_xpath_experimental.dtd\">\n");
+            bw.write("<suppressions>\n");
+            bw.write("   <suppress-xpath\n");
+            bw.write("       checks=\"");
+            bw.write(checkName);
+            bw.write("\"\n");
+            bw.write("       query=\"");
+            bw.write(xpathQueries.stream().collect(Collectors.joining(DELIMITER)));
```

Comment:
why we join all xpath queries ?

Is it better to test each query separately ? to make sure that all that we generated could suppress violation.

Reply:
Yes, we need to test generated queries one by one.

Reply:
@timurt  , what is status on this point? Is there any good reason?
Please keep replying to each point, it help a lot.